validate-browsing-context-id

manuelsblanco opened the pull request:
Enhanced the BrowsingContext Constructor to Include Additional Validation for 'id' Parameter

**Hello Selenium community! Your contribution matters and well-described pull requests make reviews and merges efficient. Thank you for your effort!**

Before submitting your pull request, please review our [contributing guidelines](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md). Keeping pull requests concise and focused helps reviewers.

<!--- Provide a general summary of your changes in the Title above -->

### Description
In this pull request, I've enhanced the `BrowsingContext` constructor to include additional validation for the 'id' parameter. The objective is to ensure that the 'id' parameter is not only non-null but also non-empty, providing an extra layer of validation. Additionally, I've introduced a validation step to confirm that the WebDriver instance supports the BiDi protocol before retrieving the BiDi instance.

### Motivation and Context
This enhancement was motivated by the need to strengthen the validation process for the 'id' parameter in the `BrowsingContext` constructor. By introducing a check for empty string values, we ensure that the 'id' parameter is not only non-null but also contains meaningful content. Additionally, verifying the support for the BiDi protocol in the WebDriver instance helps prevent compatibility issues.

### Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing guidelines](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [] I have added tests to cover my changes.
- [x] All new and existing tests passed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/selenium?pullRequest=12678) <br/>All committers have signed the CLA.

diemol commented:
@pujagani, does this make sense?

manuelsblanco commented:
**Description**

In this pull request, I've introduced the LogEntryUtils interface to enhance code reusability and centralize JSON deserialization logic for log entries. This interface provides a common method, fromJson, which simplifies the creation of log entry objects from JSON input.

The primary motivation behind this change is to eliminate code duplication in the GenericLogEntry and JavascriptLogEntry classes. The fromJson method reduces redundant code and improves maintainability. Furthermore, I've added a new test class, LogEntryUtilsTest, to thoroughly test the functionality of the LogEntryUtils interface.

**Motivation and Context**

This enhancement is motivated by the need to improve code maintainability and reduce redundancy in the codebase. By centralizing JSON deserialization logic in the LogEntryUtils interface, we ensure that the same parsing process is consistently applied to all log entry types. This change also promotes better testing practices by allowing us to easily test the deserialization process using the LogEntryUtilsTest class.

**Types of Changes**

- [x] New feature, a new interface to avoid duplicate code (non-breaking change which adds functionality)

**Checklist**

- [x] I have read the contributing guidelines.
- [] My change does not require a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<img width="1638" alt="image" src="https://github.com/SeleniumHQ/selenium/assets/30122785/982305ec-01a4-4c0e-b0e4-34972eb517a2">

manuelsblanco commented:
Hi @diemol,

I noticed that @pujagani is currently out of the office. I have added new code to the PR. If you'd like to discuss it, you can reach me on Slack.

Thanks in advance.

pujagani commented on java/src/org/openqa/selenium/bidi/browsingcontext/BrowsingContext.java:
```diff
@@ -77,6 +77,10 @@ public BrowsingContext(WebDriver driver, String id) {
       throw new IllegalArgumentException("WebDriver instance must support BiDi protocol");
     }
 
+    if (id.isEmpty()) {
```
It might be a good idea to use "Require.precondition" to check for null and empty string. 
https://github.com/SeleniumHQ/selenium/blob/e7c60f5e20e0f9c00bb451884b2bda03b093d92c/java/src/org/openqa/selenium/internal/Require.java#L52

pujagani commented:
@manuelsblanco  I appreciate your help here. I agree with the motivation stated. However, I see two unrelated changes (motivation-wise) in the same PR. Can you please help split it up? Even if the change is small. It will be easier to review and land them independently. Thank you!

manuelsblanco commented:
> I appreciate your help here. I agree with the motivation stated. However, I see two unrelated changes (motivation-wise) in the same PR. Can you please help split it up? Even if the change is small. It will be easier to review and land them independently. Thank you!

Sure thing, I'll do this split as soon as possible.

manuelsblanco commented:
Hi @pujagani, I have just implemented your suggestion regarding the use of Require.Precondition. Please let me know if there is anything else I can do.

pujagani commented:
Thank you for incorporating the changes. Can you please help split up the PR as well? Thank you!

manuelsblanco commented:
This PR was split in two, we can close this one now.
